## Assignment 3 - Atsu Atakpa

### Question 1
Compare your given name with your nickname (if you don’t have a nickname, invent one for this assignment) by answering the following questions:

a.	What is the edit distance between your nickname and your given name? <br>
b.	What is the percentage string match between your nickname and your given name?

Show your work for both calculations.

### Answer 


Edit distance measures how near or far a word is from our input word.

### My edit distance for my name "Atsu" and my nick name "tsu-tsu" is 4.

In [1]:
from __future__ import division
#this calculates edit distance not levenstein edit distance
word1="Atsu"

word2="tsu-tsu"

len_1=len(word1)

len_2=len(word2)

x =[[0]*(len_2+1) for _ in range(len_1+1)]#the matrix whose last element ->edit distance

for i in range(0,len_1+1): #initialization of base case values

    x[i][0]=i
for j in range(0,len_2+1):

    x[0][j]=j
for i in range (1,len_1+1):

    for j in range(1,len_2+1):

        if word1[i-1]==word2[j-1]:
            x[i][j] = x[i-1][j-1] 

        else :
            x[i][j]= min(x[i][j-1],x[i-1][j],x[i-1][j-1])+1

print ("The edit distance for my name 'Atsu' to my nickname 'tsu-tsu' is: %.2f" % x[i][j])

The edit distance for my name 'Atsu' to my nickname 'tsu-tsu' is: 4.00


#### b.	What is the percentage string match between your nickname and your given name?

In [2]:
print("First name length: %.2f" %len_1)
print("Second name length: %.2f" %len_2)
print("Sum of length: %.2f" %(len_1 + len_2))
print("percentage string match between your nickname and your given name: %.2f" %(x[i][j] / (len_1 + len_2) * 100))

First name length: 4.00
Second name length: 7.00
Sum of length: 11.00
percentage string match between your nickname and your given name: 36.36


#### My percentage string match between my nick name and given name is 36.36%

Another example of calculating the edit distance

In [3]:
def LD(s, t):
    if s == "":
        return len(t)
    if t == "":
        return len(s)
    if s[-1] == t[-1]:
        cost = 0
    else:
        cost = 1
       
    res = min([LD(s[:-1], t)+1,
               LD(s, t[:-1])+1, 
               LD(s[:-1], t[:-1]) + cost])
    return res
print(LD("Atsu", "tsu-tsu"))

4


### Question 2

Find a friend (or family member or classmate) who you know has read a certain book. Without your friend knowing, copy the first two sentences of that book. <br> Now rewrite the words from those sentences, excluding stop words. <br> Now tell your friend to guess which book the words are from by reading them just that list of words.<br> Did you friend correctly guess the book on the first try?<br> What did he or she guess?<br> Explain why you think you friend either was or was not able to guess the book from hearing the list of words. 

### Answer

* The text is from the book "The fox and the hound" by Daniel P. Mannix

My friend was able to correctly guess the book. One of the main reasons why she could guess the book was because it is a common book. She read it a couple of times when she was younger. However, the output of the words also gave it away. Output such as
   'The',
   'big',
   'half-bred',
   'bloodhound'
Made it much easier for her. The guess wasn't instant but she got it. If this exercise was conducted on a book she has probably read only once, I'm not sure the outcome will be the same.

In [4]:
import nltk
import re
import string
from pprint import pprint

In [5]:
corpus =["The big half-bred bloodhound lay in his barrel kennel and dreamed he was a deer hunting. Of all the quarry he had ever trailed, deer was the hound's favorite, although it was usually strictly forbidden. He had been whipped, beaten with a club, and even hit in the flanks by bird shot from this crime; yet once, after an interminable journey to a faraway place, he had been allowed to track a deer."]

In [6]:
print corpus

["The big half-bred bloodhound lay in his barrel kennel and dreamed he was a deer hunting. Of all the quarry he had ever trailed, deer was the hound's favorite, although it was usually strictly forbidden. He had been whipped, beaten with a club, and even hit in the flanks by bird shot from this crime; yet once, after an interminable journey to a faraway place, he had been allowed to track a deer."]


In [7]:
def tokenize_text(text):
    sentences = nltk.sent_tokenize(text)
    word_tokens = [nltk.word_tokenize(sentence) for sentence in sentences]
    return word_tokens

In [8]:
token_list = [tokenize_text(text) for text in corpus]
pprint (token_list)
print


[[['The',
   'big',
   'half-bred',
   'bloodhound',
   'lay',
   'in',
   'his',
   'barrel',
   'kennel',
   'and',
   'dreamed',
   'he',
   'was',
   'a',
   'deer',
   'hunting',
   '.'],
  ['Of',
   'all',
   'the',
   'quarry',
   'he',
   'had',
   'ever',
   'trailed',
   ',',
   'deer',
   'was',
   'the',
   'hound',
   "'s",
   'favorite',
   ',',
   'although',
   'it',
   'was',
   'usually',
   'strictly',
   'forbidden',
   '.'],
  ['He',
   'had',
   'been',
   'whipped',
   ',',
   'beaten',
   'with',
   'a',
   'club',
   ',',
   'and',
   'even',
   'hit',
   'in',
   'the',
   'flanks',
   'by',
   'bird',
   'shot',
   'from',
   'this',
   'crime',
   ';',
   'yet',
   'once',
   ',',
   'after',
   'an',
   'interminable',
   'journey',
   'to',
   'a',
   'faraway',
   'place',
   ',',
   'he',
   'had',
   'been',
   'allowed',
   'to',
   'track',
   'a',
   'deer',
   '.']]]



In [9]:
def remove_characters_after_tokenization(tokens):
    pattern = re.compile('[{}]'.format(re.escape(string.punctuation)))
    filtered_tokens = filter(None, [pattern.sub('', token) for token in tokens])
    return filtered_tokens

In [10]:
filtered_list_1 =  [filter(None,[remove_characters_after_tokenization(tokens) 
                                for tokens in sentence_tokens]) 
                    for sentence_tokens in token_list]
print filtered_list_1
print 

[[['The', 'big', 'halfbred', 'bloodhound', 'lay', 'in', 'his', 'barrel', 'kennel', 'and', 'dreamed', 'he', 'was', 'a', 'deer', 'hunting'], ['Of', 'all', 'the', 'quarry', 'he', 'had', 'ever', 'trailed', 'deer', 'was', 'the', 'hound', 's', 'favorite', 'although', 'it', 'was', 'usually', 'strictly', 'forbidden'], ['He', 'had', 'been', 'whipped', 'beaten', 'with', 'a', 'club', 'and', 'even', 'hit', 'in', 'the', 'flanks', 'by', 'bird', 'shot', 'from', 'this', 'crime', 'yet', 'once', 'after', 'an', 'interminable', 'journey', 'to', 'a', 'faraway', 'place', 'he', 'had', 'been', 'allowed', 'to', 'track', 'a', 'deer']]]



In [11]:
def remove_characters_before_tokenization(sentence,
                                          keep_apostrophes=False):
    sentence = sentence.strip()
    if keep_apostrophes:
        PATTERN = r'[?|$|&|*|%|@|(|)|~]'
        filtered_sentence = re.sub(PATTERN, r'', sentence)
    else:
        PATTERN = r'[^a-zA-Z0-9 ]'
        filtered_sentence = re.sub(PATTERN, r'', sentence)
    return filtered_sentence

In [12]:
filtered_list_2 = [remove_characters_before_tokenization(sentence) 
                    for sentence in corpus]    
print filtered_list_2
print 

['The big halfbred bloodhound lay in his barrel kennel and dreamed he was a deer hunting Of all the quarry he had ever trailed deer was the hounds favorite although it was usually strictly forbidden He had been whipped beaten with a club and even hit in the flanks by bird shot from this crime yet once after an interminable journey to a faraway place he had been allowed to track a deer']



In [13]:
cleaned_corpus = [remove_characters_before_tokenization(sentence, keep_apostrophes=True) 
                  for sentence in corpus]
print cleaned_corpus

["The big half-bred bloodhound lay in his barrel kennel and dreamed he was a deer hunting. Of all the quarry he had ever trailed, deer was the hound's favorite, although it was usually strictly forbidden. He had been whipped, beaten with a club, and even hit in the flanks by bird shot from this crime; yet once, after an interminable journey to a faraway place, he had been allowed to track a deer."]


In [14]:
CONTRACTION_MAP = {
"ain't": "is not",
"aren't": "are not",
"can't": "cannot",
"can't've": "cannot have",
"'cause": "because",
"could've": "could have",
"couldn't": "could not",
"couldn't've": "could not have",
"didn't": "did not",
"doesn't": "does not",
"don't": "do not",
"hadn't": "had not",
"hadn't've": "had not have",
"hasn't": "has not",
"haven't": "have not",
"he'd": "he would",
"he'd've": "he would have",
"he'll": "he will",
"he'll've": "he he will have",
"he's": "he is",
"how'd": "how did",
"how'd'y": "how do you",
"how'll": "how will",
"how's": "how is",
"I'd": "I would",
"I'd've": "I would have",
"I'll": "I will",
"I'll've": "I will have",
"I'm": "I am",
"I've": "I have",
"i'd": "i would",
"i'd've": "i would have",
"i'll": "i will",
"i'll've": "i will have",
"i'm": "i am",
"i've": "i have",
"isn't": "is not",
"it'd": "it would",
"it'd've": "it would have",
"it'll": "it will",
"it'll've": "it will have",
"it's": "it is",
"let's": "let us",
"ma'am": "madam",
"mayn't": "may not",
"might've": "might have",
"mightn't": "might not",
"mightn't've": "might not have",
"must've": "must have",
"mustn't": "must not",
"mustn't've": "must not have",
"needn't": "need not",
"needn't've": "need not have",
"o'clock": "of the clock",
"oughtn't": "ought not",
"oughtn't've": "ought not have",
"shan't": "shall not",
"sha'n't": "shall not",
"shan't've": "shall not have",
"she'd": "she would",
"she'd've": "she would have",
"she'll": "she will",
"she'll've": "she will have",
"she's": "she is",
"should've": "should have",
"shouldn't": "should not",
"shouldn't've": "should not have",
"so've": "so have",
"so's": "so as",
"that'd": "that would",
"that'd've": "that would have",
"that's": "that is",
"there'd": "there would",
"there'd've": "there would have",
"there's": "there is",
"they'd": "they would",
"they'd've": "they would have",
"they'll": "they will",
"they'll've": "they will have",
"they're": "they are",
"they've": "they have",
"to've": "to have",
"wasn't": "was not",
"we'd": "we would",
"we'd've": "we would have",
"we'll": "we will",
"we'll've": "we will have",
"we're": "we are",
"we've": "we have",
"weren't": "were not",
"what'll": "what will",
"what'll've": "what will have",
"what're": "what are",
"what's": "what is",
"what've": "what have",
"when's": "when is",
"when've": "when have",
"where'd": "where did",
"where's": "where is",
"where've": "where have",
"who'll": "who will",
"who'll've": "who will have",
"who's": "who is",
"who've": "who have",
"why's": "why is",
"why've": "why have",
"will've": "will have",
"won't": "will not",
"won't've": "will not have",
"would've": "would have",
"wouldn't": "would not",
"wouldn't've": "would not have",
"y'all": "you all",
"y'all'd": "you all would",
"y'all'd've": "you all would have",
"y'all're": "you all are",
"y'all've": "you all have",
"you'd": "you would",
"you'd've": "you would have",
"you'll": "you will",
"you'll've": "you will have",
"you're": "you are",
"you've": "you have"
}

In [15]:
#from contractions import CONTRACTION_MAP

def expand_contractions(sentence, contraction_mapping):
    
    contractions_pattern = re.compile('({})'.format('|'.join(contraction_mapping.keys())), 
                                      flags=re.IGNORECASE|re.DOTALL)
    def expand_match(contraction):
        match = contraction.group(0)
        first_char = match[0]
        expanded_contraction = contraction_mapping.get(match)\
                                if contraction_mapping.get(match)\
                                else contraction_mapping.get(match.lower())                       
        expanded_contraction = first_char+expanded_contraction[1:]
        return expanded_contraction
        
    expanded_sentence = contractions_pattern.sub(expand_match, sentence)
    return expanded_sentence
    
expanded_corpus = [expand_contractions(sentence, CONTRACTION_MAP) 
                    for sentence in cleaned_corpus]    
print expanded_corpus
print

["The big half-bred bloodhound lay in his barrel kennel and dreamed he was a deer hunting. Of all the quarry he had ever trailed, deer was the hound's favorite, although it was usually strictly forbidden. He had been whipped, beaten with a club, and even hit in the flanks by bird shot from this crime; yet once, after an interminable journey to a faraway place, he had been allowed to track a deer."]



In [16]:
# case conversion    
print corpus[0].lower()
print corpus[0].upper()

the big half-bred bloodhound lay in his barrel kennel and dreamed he was a deer hunting. of all the quarry he had ever trailed, deer was the hound's favorite, although it was usually strictly forbidden. he had been whipped, beaten with a club, and even hit in the flanks by bird shot from this crime; yet once, after an interminable journey to a faraway place, he had been allowed to track a deer.
THE BIG HALF-BRED BLOODHOUND LAY IN HIS BARREL KENNEL AND DREAMED HE WAS A DEER HUNTING. OF ALL THE QUARRY HE HAD EVER TRAILED, DEER WAS THE HOUND'S FAVORITE, ALTHOUGH IT WAS USUALLY STRICTLY FORBIDDEN. HE HAD BEEN WHIPPED, BEATEN WITH A CLUB, AND EVEN HIT IN THE FLANKS BY BIRD SHOT FROM THIS CRIME; YET ONCE, AFTER AN INTERMINABLE JOURNEY TO A FARAWAY PLACE, HE HAD BEEN ALLOWED TO TRACK A DEER.


In [17]:
# removing stopwords
def remove_stopwords(tokens):
    stopword_list = nltk.corpus.stopwords.words('english')
    filtered_tokens = [token for token in tokens if token not in stopword_list]
    return filtered_tokens
    
expanded_corpus_tokens = [tokenize_text(text)
                          for text in expanded_corpus]    
filtered_list_3 =  [[remove_stopwords(tokens) 
                        for tokens in sentence_tokens] 
                        for sentence_tokens in expanded_corpus_tokens]
print filtered_list_3
print 

[[['The', 'big', 'half-bred', 'bloodhound', 'lay', 'barrel', 'kennel', 'dreamed', 'deer', 'hunting', '.'], ['Of', 'quarry', 'ever', 'trailed', ',', 'deer', 'hound', "'s", 'favorite', ',', 'although', 'usually', 'strictly', 'forbidden', '.'], ['He', 'whipped', ',', 'beaten', 'club', ',', 'even', 'hit', 'flanks', 'bird', 'shot', 'crime', ';', 'yet', ',', 'interminable', 'journey', 'faraway', 'place', ',', 'allowed', 'track', 'deer', '.']]]



### Question 3

3.	Run one of the stemmers available in Python. Run the same two sentences from question 2 above through the stemmer and show the results. How many of the outputted stems are valid morphological roots of the corresponding words? Express this answer as a percentage.

### Answer 3

In [18]:
from nltk.stem import PorterStemmer
ps = PorterStemmer()

In [19]:
type(filtered_list_3)

list

In [20]:
print filtered_list_3[0][0]

['The', 'big', 'half-bred', 'bloodhound', 'lay', 'barrel', 'kennel', 'dreamed', 'deer', 'hunting', '.']


In [21]:
print filtered_list_3[0][1]

['Of', 'quarry', 'ever', 'trailed', ',', 'deer', 'hound', "'s", 'favorite', ',', 'although', 'usually', 'strictly', 'forbidden', '.']


In [22]:
sen1=((word.lower() for word in filtered_list_3[0][0] if word.isalpha())) # new dataframe that cleans up capitlizations, and aplha

In [23]:
for w in sen1:
    print(ps.stem(w))

the
big
bloodhound
lay
barrel
kennel
dream
deer
hunt


In [24]:
sen2=((word.lower() for word in filtered_list_3[0][1] if word.isalpha())) # new dataframe that cleans up capitlizations, and aplha

In [25]:
sen2

<generator object <genexpr> at 0x000000000A5CCB88>

In [26]:
print sen2

<generator object <genexpr> at 0x000000000A5CCB88>


In [27]:
for w in sen2:
    print(ps.stem(w))

of
quarri
ever
trail
deer
hound
favorit
although
usual
strictli
forbidden


SOURCE:https://stackoverflow.com/questions/2460177/edit-distance-in-python
https://www.python-course.eu/levenshtein_distance.php
https://stackoverflow.com/questions/2460177/edit-distance-in-python